File: apps/android/app/src/main/kotlin/jiamin/chen/orangecloud/core/network/CfApiResponse.kt
package jiamin.chen.orangecloud.core.network

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Cloudflare API 通用信封 { result, success, errors, messages, result_info }。
 * 与 iOS Models/CFAPIResponse.swift 对应；这里用单一信封承载可选 result_info（列表端点才有）。
 */
@Serializable
data class CfEnvelope<T>(
    val result: T? = null,
    val success: Boolean = false,
    val errors: List<CfApiError> = emptyList(),
    val messages: List<CfMessage> = emptyList(),
    @SerialName("result_info") val resultInfo: ResultInfo? = null,
)

@Serializable
data class CfApiError(val code: Int = 0, val message: String = "")

@Serializable
data class CfMessage(val code: Int = 0, val message: String = "")

@Serializable
data class ResultInfo(
    // 页码分页（Zone/DNS 等）
    val page: Int? = null,
    @SerialName("per_page") val perPage: Int? = null,
    @SerialName("total_pages") val totalPages: Int? = null,
    val count: Int? = null,
    @SerialName("total_count") val totalCount: Int? = null,
    // 游标分页（R2 对象、KV keys 等）
    val cursor: String? = null,
    @SerialName("is_truncated") val isTruncated: Boolean? = null,
    // R2 带 delimiter 列举时的「文件夹」前缀。实测 CF 返回的键就叫 delimited
    // （不是 delimited_prefixes），与 iOS CFAPIResponse.ResultInfo.delimited 一致。
    @SerialName("delimited") val delimitedPrefixes: List<String>? = null,
)

/** 列表端点的解码结果：数据 + 分页信息 */
data class Paged<T>(val items: List<T>, val info: ResultInfo?)

// MARK: - GraphQL（分析端点，信封是 {data, errors}，GraphQL 错误时 HTTP 仍 200）

@Serializable
data class GraphQLRequest<V>(val query: String, val variables: V)

@Serializable
data class GraphQLResponse<D>(
    val data: D? = null,
    val errors: List<GraphQLError> = emptyList(),
)

@Serializable
data class GraphQLError(val message: String = "") {
    /**
     * 账户级数据集未授权。CF 的 GraphQL 不在 extensions 里给结构化码（安卓侧信封没解 extensions），
     * 只能按文案识别——与 DashboardViewModel 早先的判定口径一致，收拢到这里供各处复用。
     */
    val isAuthz: Boolean
        get() = message.lowercase().let { m ->
            "authz" in m || "not authorized" in m || "unauthorized" in m ||
                "permission" in m || "authentication" in m
        }
}

File: apps/ios/Orange Cloud/Orange Cloud/Models/CFAPIResponse.swift
//
//  CFAPIResponse.swift
//  Orange Cloud
//
//  Cloudflare API 通用响应包装 { result, success, errors, messages }
//

import Foundation

nonisolated struct CFAPIResponse<T: Codable & Sendable>: Codable, Sendable {
    let result:   T?
    let success:  Bool
    let errors:   [CFAPIError]
    let messages: [CFAPIMessage]?
    /// 少数对象型 result 的端点也带分页信息（R2 桶列表的 cursor）
    let resultInfo: ResultInfo?

    enum CodingKeys: String, CodingKey {
        case result, success, errors, messages
        case resultInfo = "result_info"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        result   = try c.decodeIfPresent(T.self, forKey: .result)
        success  = try c.decode(Bool.self, forKey: .success)
        resultInfo = try? c.decodeIfPresent(ResultInfo.self, forKey: .resultInfo)
        // 部分端点（如 workers/domains）回 errors:null，宽容降级为空数组
        errors   = (try? c.decode([CFAPIError].self, forKey: .errors)) ?? []
        // messages 只是附加提示，形态不合预期（如 Observability 端点只回 {message} 没有 code）
        // 绝不能拖垮整条响应（Sentry APPLE-IOS-BQ）
        messages = try? c.decodeIfPresent([CFAPIMessage].self, forKey: .messages)
    }
}

nonisolated struct CFAPIResponseArray<T: Codable & Sendable>: Codable, Sendable {
    let result:     [T]?
    let success:    Bool
    let errors:     [CFAPIError]
    let resultInfo: ResultInfo?

    enum CodingKeys: String, CodingKey {
        case result, success, errors
        case resultInfo = "result_info"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        result     = try c.decodeIfPresent([T].self, forKey: .result)
        success    = try c.decode(Bool.self, forKey: .success)
        // 部分端点回 errors:null，宽容降级为空数组
        errors     = (try? c.decode([CFAPIError].self, forKey: .errors)) ?? []
        resultInfo = try c.decodeIfPresent(ResultInfo.self, forKey: .resultInfo)
    }
}

nonisolated struct CFAPIError: Codable, Sendable {
    let code:    Int
    let message: String
    /// 2026-08-20 起 CF 在 4xx（尤其 403）的错误体里带上该端点所需角色/权限的文档地址。
    /// 老端点与旧响应可能没有，故为可选。
    let documentationURL: String?
    /// 出错字段的 JSON Pointer（如 "/body/rules/0/action"），字段级校验失败时才有。
    let source: CFAPIErrorSource?

    enum CodingKeys: String, CodingKey {
        case code, message, source
        case documentationURL = "documentation_url"
    }

    /// 两个新字段一律宽容解码：它们只是排查用的附加信息，形态不合预期时也绝不能
    /// 拖垮整条错误的解码——外层 `errors` 解不出会整体降级为空数组，把 code/message
    /// 一起丢掉，错误信息反而不如加字段之前。
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        code    = try c.decode(Int.self, forKey: .code)
        message = try c.decode(String.self, forKey: .message)
        documentationURL = try? c.decodeIfPresent(String.self, forKey: .documentationURL)
        source           = try? c.decodeIfPresent(CFAPIErrorSource.self, forKey: .source)
    }
}

nonisolated struct CFAPIErrorSource: Codable, Sendable {
    let pointer: String?
}

nonisolated struct CFAPIMessage: Codable, Sendable {
    /// 部分端点（Workers Observability）的 messages 项没有 code
    let code:    Int?
    let message: String
}

nonisolated struct ResultInfo: Codable, Sendable {
    // 页码分页（Zone/DNS 等）
    let page:       Int?
    let perPage:    Int?
    let totalPages: Int?
    let count:      Int?
    let totalCount: Int?
    // 游标分页（R2 对象、KV keys 等）
    let cursor:      String?
    // R2 list 传 delimiter 时回的「折叠前缀」（即子文件夹），key 就叫 delimited
    let delimited:   [String]?
    let isTruncated: Bool?

    enum CodingKeys: String, CodingKey {
        case page, count, cursor, delimited
        case perPage     = "per_page"
        case totalPages  = "total_pages"
        case totalCount  = "total_count"
        case isTruncated = "is_truncated"
    }
}

// 用于 DELETE 等只关心 success 的请求
nonisolated struct EmptyResponse: Codable, Sendable {}

extension CFAPIResponse {
    func toAPIError() -> APIError {
        let err = errors.first
        return .cloudflareError(code: err?.code ?? 0,
                                message: err?.message ?? String(localized: "未知错误"),
                                documentationURL: err?.documentationURL)
    }
}

extension CFAPIResponseArray {
    func toAPIError() -> APIError {
        let err = errors.first
        return .cloudflareError(code: err?.code ?? 0,
                                message: err?.message ?? String(localized: "未知错误"),
                                documentationURL: err?.documentationURL)
    }
}
